proxy/ParentYamlConfig.cc

```diff
@@ -0,0 +1,224 @@
+    p.parent = node["parent"].as<std::string>();
+  }
+  if (node["secondary_parent"]) {
+    p.secondary_parent = node["secondary_parent"].as<std::string>();
+  }
+  if (node["parent_retry"]) {
+    p.parent_retry = node["parent_retry"].as<std::string>();
+  }
+  if (node["unavailable_server_retry_responses"]) {
+    p.unavailable_server_retry_responses = node["unavailable_server_retry_responses"].as<std::string>();
+  }
+  if (node["round_robin"]) {
+    p.round_robin = node["round_robin"].as<std::string>();
+  }
+  if (node["qstring"]) {
+    p.qstring = node["qstring"].as<std::string>();
+  }
+
+  // get all the int's
+  if (node["port"]) {
+    p.port = node["port"].as<int>();
+  }
+  if (node["secondary_mode"]) {
+    p.secondary_mode = node["secondary_mode"].as<int>();
+  }
+  if (node["max_simple_retries"]) {
+    p.max_simple_retries = node["max_simple_retries"].as<int>();
+  }
+  if (node["max_unavailable_server_retries"]) {
+    p.max_unavailable_server_retries = node["max_unavailable_server_retries"].as<int>();
+  }
+
+  // get all the booleans
+  if (node["internal"]) {
+    p.internal = node["internal"].as<bool>();
+  }
+  if (node["parent_is_proxy"]) {
+    p.parent_is_proxy = node["parent_is_proxy"].as<bool>();
+  }
+  if (node["go_direct"]) {
+    p.go_direct = node["go_direct"].as<bool>();
+  }
+
+  return p;
+}
+
+int
+ParentYamlConfig::Init(const std::string &content)
+{
+  YAML::Node root;
+  YAML::Node destinations;
+
+  try {
+    root = YAML::Load(content);
+
+    if (!root.IsMap()) {
+      error_msg << __FILE__ << ":" << __LINE__ << ": "
+                << "yaml parse error, expected a map at the root node.";
+      return 1;
+    }
+    destinations = root["destinations"];
+    if (!destinations.IsSequence()) {
+      error_msg << __FILE__ << ":" << __LINE__ << ": "
+                << "yaml parse error, expected a 'destinations' node of type sequence.";
+      return 1;
+    } else {
+      dest_size = destinations.size();
+    }
+    for (uint i = 0; i < dest_size; i++) {
+      parent_dest *p = new parent_dest();
+      destinations[i] >> *p;
+      if (verifyConfig(*p, i + 1)) {
+        std::unique_ptr<parent_dest> pd(p);
+        parent_configs.push_back(std::move(pd));
+      }
+    }
+  } catch (YAML::ParserException &e) {
+    error_msg << __FILE__ << ":" << __LINE__ << ": " << e.what();
+    return 1;
+  }
+
+  return 0;
+}
+
+bool
+ParentYamlConfig::verifyConfig(const parent_dest &p, uint entry)
+{
+  bool result = true;
+
+  uint dest_set = 0;
+  if (!p.dest_domain.empty()) {
+    dest_set++;
+  }
+  if (!p.dest_host.empty()) {
+    dest_set++;
+  }
+  if (!p.dest_ip.empty()) {
+    dest_set++;
+  }
+  if (!p.url_regex.empty()) {
+    dest_set++;
+  }
+
+  if (dest_set == 0 || dest_set > 1) {
+    Error("%s:%d: Skipping entry %d, invalid 'destination', exactly one of 'dest_domain', 'dest_host', 'dest_ip', or 'url_regex' "
+          "must be set",
+          __FILE__, __LINE__, entry);
```

Comment:
Nitpick:
```
    Error(__FILE__ ":%d: Skipping entry %d, invalid 'destination', exactly one of 'dest_domain', 'dest_host', 'dest_ip', or 'url_regex' "
          "must be set",
          __LINE__, entry);
```

Reply:
Note you can get a config file source line number and line position from the YAML node.

Reply:
resolved, this has be re-implemented.